### PR TITLE
DOC Document code moved from cms to framework

### PR DESCRIPTION
**Pull request by GuySartorelli**
## Issue
- https://github.com/silverstripe/silverstripe-cms/issues/2947

### CODE REVIEW

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/How_Tos/Customise_CMS_Tree.md
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/How_Tos/Customise_CMS_Tree.md:1]: Note the rest of the docs on this page will be retouched in https://github.com/silverstripe/silverstripe-cms/issues/2951

File: en/08_Changelogs/6.0.0.md [hunk 1]
```diff
@@ -617,6 +617,14 @@ As a result of this change, to following classes are now direct subclasses of `A
 
 The `tree_class` configuration property on `LeftAndMain` and its subclasses has be renamed to [`model_class`](api:SilverStripe\Admin\LeftAndMain->model_class). This is used in methods like [`getRecord()`](api:SilverStripe\Admin\LeftAndMain::getRecord()) to get a record of the correct class.
 
+#### Effects of this refactor in other classes
+
+Some classes outside of the `LeftAndMain` class hierarchy have also been affected by the refactoring:
+
+- The `SiteTree.description` configuration property has been renamed to [`class_description`](api:SilverStripe\ORM\DataObject->class_description).
+    This configuration has been added to `DataObject` along with the corresponding [`classDescription()`](api:SilverStripe\ORM\DataObject::classDescription()) and [`i18n_classDescription()`](api:SilverStripe\ORM\DataObject::i18n_classDescription()) methods.
+- The [`Hierarchy`](api:SilverStripe\ORM\Hierarchy\Hierarchy) extension now has a bunch of configuration and methods which used to be exclusive to `SiteTree`.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md:639]: Part of https://github.com/silverstripe/silverstripe-cms/issues/2951 will be updating the CMS Tree docs which will explain more about `Hierarchy` and how it's used - when that happens this note will link to that page.

File: en/08_Changelogs/6.0.0.md [hunk 2]
```diff
@@ -675,6 +683,7 @@ You can change the level of entropy required by passing one of the valid [minSco
 - The [`UrlField`](api:SilverStripe\Forms\UrlField) class has some new API for setting which protocols are allowed for valid URLs.
 - The [`EmailField`](api:SilverStripe\Forms\EmailField) class now uses `symfony/validator` to handle its validation logic, where previously this was validated with a custom regex.
 - [`ArrayData`](api:SilverStripe\Model\ArrayData) can now be serialised using [`json_encode()`](https://www.php.net/manual/en/function.json-encode.php).
+- There's a new [`ModelData::getStatusFlags()`](api:SilverStripe\Model\ModelData::getStatusFlags()) method for status flags to display in the CMS. This can be used for example to show what locale the data is in, or what versioned stage it's in, etc. This is used in a few limited places in the CMS but is likely to be used more in future minor releases.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md]: Thought about giving this its own section but given we're not really using it in many places right now it got downgraded to an "other new feature".

File: en/08_Changelogs/6.0.0.md [hunk 3]
```diff
@@ -871,8 +880,8 @@ As part of these changes [`ArrayList::find()`](api:SilverStripe\Model\List\Array
...
-- [`DataExtension`](api:SilverStripe\ORM\DataExtension), [`SiteTreeExtension`](api:SilverStripe\CMS\Model\SiteTreeExtension), and [`LeftAndMainExtension`](api:SilverStripe\Admin\LeftAndMainExtension) have been removed. If you subclass any of these classes, you must now subclass [`Extension`](api:SilverStripe\Core\Extension) instead.
+- `DNADesign\Elemental\Models\BaseElement::getDescription()` and the corresponding `DNADesign\Elemental\Models\BaseElement.description` configuration property have been removed. If you were using either of these in your custom elemental blocks, either set the new [`class_description`](api:SilverStripe\ORM\DataObject->class_description) configuration property or override one of the [`i18n_classDescription()`](api:SilverStripe\ORM\DataObject::i18n_classDescription()) or [`getTypeNice()`](api:DNADesign\Elemental\Models\BaseElement::getTypeNice()) methods instead.
+- `SilverStripe\ORM\DataExtension`, `SilverStripe\CMS\Model\SiteTreeExtension`, and `SilverStripe\Admin\LeftAndMainExtension` have been removed. If you subclass any of these classes, you must now subclass [`Extension`](api:SilverStripe\Core\Extension) directly instead.
```
Comment (ChloeSartorelli) [inline en/08_Changelogs/6.0.0.md:909]: Removing links is unrelated refactoring - I noticed that these are links to classes that have been removed, and therefore the links won't work.

Mixed in here is changes which _are_ related to the issue - mentioning the new `class_description` config and `classDescription()` methods can be used to replace the old `description` config and `getDescription()` method.

File: en/08_Changelogs/6.0.0.md (hunk 2 above)
Comment (emteknetnz) [reply en/08_Changelogs/6.0.0.md]: Won't this now be applied to any versioned DataObject in a GridField?

If not, we should include some instructions (not in the changelog, but elsewhere) about how to add these since it's a new feature? e.g. https://github.com/silverstripe/silverstripe-frameworktest/pull/212/files

**Review by emteknetnz — CHANGES_REQUESTED** [review]

**Comment by GuySartorelli** [comment]
Yup, forgot to update this. The whole "actually using it" thing was gonna be done in minors originally. Will update this now.

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/How_Tos/Customise_CMS_Tree.md
Comment (ChloeSartorelli) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/How_Tos/Customise_CMS_Tree.md:1]: Moved content from here into a new page, since this isn't specific to `SiteTree` anymore.

File: en/_images/page_node_normal.png
Comment (ChloeSartorelli) [inline en/_images/page_node_normal.png:1]: Not used anymore in favour of the new screenshots

Comment (emteknetnz) [reply en/_images/page_node_normal.png:1]: status-flag.png shouldn't include the fluent icons since most install won't have those, so this may confuse people why their site does not have them

status-flags-custom.png has misaligned text in the badges

File: en/08_Changelogs/6.0.0.md [hunk 4]
```diff
@@ -679,6 +688,18 @@ This has been changed to use the [`PasswordStrength` constraint](https://symfony
...
 
+### Status flags in the CMS
+
+In CMS 5 the CMS showed status flags for the [`SiteTree`](api:SilverStripe\CMS\Model\SiteTree) class inside the `/admin/pages` section, and occasionally for other models in grid fields - but this was inconsistent and was done in a variety of different ways depending on what was adding the flags and where they were displayed.
+
+We've standardised this in the new [`ModelData::getStatusFlags()`](api:SilverStripe\Model\ModelData::getStatusFlags()) method to define the flags, and [`ModelData::getStatusFlagMarkup()`](api:SilverStripe\Model\ModelData::getStatusFlagMarkup()) to build the HTML markup for them. This means that status flags can be displayed for *any* model in the CMS.
+
+This is already used to show what locale the data is in for models localised using [`tractorcow/silverstripe-fluent`](https://github.com/tractorcow-farm/silverstripe-fluent/), and what versioned stage it's in for models using the [`Versioned`](api:SilverStripe\Versioned\Versioned) extension.
+
+Status flags are displayed in breadcrumbs at the top of edit forms in the CMS, in the site tree for `CMSMain`, for each row in a grid field, and in [``]() and [``]().
```
Comment (emteknetnz) [inline en/08_Changelogs/6.0.0.md]: > and in [``]() and [``]()

Looks like you've forgotten to fill in some text and links here

File: en/08_Changelogs/6.0.0.md (hunk 1 above)
Comment (emteknetnz) [reply en/08_Changelogs/6.0.0.md:639]: This note will link to that page? I'm not sure what this means? Does that mean you need to add something else to this PR?

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    // ...
+    private static array $db = [
+        'IsPublishScheduled' => 'Boolean',
+    ];
+
+    public function getStatusFlags(bool $cached = true): array
+    {
+        $this->beforeExtending('updateStatusFlags', function (array &$flags) {
+            if ($this->IsPublishScheduled) {
+                $flags['scheduledtopublish'] = _t(__CLASS__ . 'SCHEDULE_PUBLISH', 'Scheduled To publish');
```
Comment (emteknetnz) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md]:
```suggestion
    private static array $db = [
        'MyColour' => 'Varchar',
    ];

    public function getStatusFlags(bool $cached = true): array
    {
        $this->beforeExtending('updateStatusFlags', function (array &$flags) {
            if (in_array($this->MyColour, ['Red', 'Green'])) {
                $flags['mycolour'] = $this->MyColour;
```

I think it's better to just use arbitary example data rather than using "IsPublishScheduled" which kind of implies there's "scheduled publishing" functionality somewhere in the CMS, and I think we should just avoid "publishing" in general here as that's related to Versioning, and the point of this is we're abstracting out something that's traditionally been there mostly just there to provide versioned status badges

Also I don't think including the _t() function here is helpful as that's another concept all together, and I think the example should only be demonstrating the primary things it's trying to demonstrate

The simpler the example is the easier it is to follow

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+### Including additional context
+
+Sometimes the text of a status flag won't convey all the information you want it to, since it is so short.
+
+You can include additional context which will be displayed when hovering over your status flags. This is done using an associative array where the main text of the flag is in the `text` key, and the additional information is in the `title` key:
+
+```php
+namespace App\Model;
```
Comment (emteknetnz) [inline en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md:84]: I think this example should be changed like I've recommended above i.e. just use "MyColour" and don't use _t()

**Review by emteknetnz — CHANGES_REQUESTED** [review]

File: en/08_Changelogs/6.0.0.md (hunk 1 above)
Comment (ChloeSartorelli) [reply en/08_Changelogs/6.0.0.md:639]: When https://github.com/silverstripe/silverstripe-cms/issues/2951 is done, it will include new documentation about how `Hierarchy` works. I will be adding a link to that new page to this bullet point (something like "see <new section> for more details")

Nothing to add for this PR. It's just a note about why this bullet point doesn't _currently_ link away to more details (because there isn't appropriate docs for that right now) and seems a bit vague on its own.

File: en/08_Changelogs/6.0.0.md (hunk 4 above)
Comment (ChloeSartorelli) [reply en/08_Changelogs/6.0.0.md]: Oops, will fix this.

File: en/_images/page_node_normal.png
Comment (ChloeSartorelli) [reply en/_images/page_node_normal.png:1]: > status-flag.png shouldn't include the fluent icons since most install won't have those, so this may confuse people why their site does not have them

I've explicitly mentioned fluent as a thing that can add status flags, and pointed out that those status flags specifically come from there. Fluent is a supported module. I don't see anything wrong with including them in this screenshot and I don't think it's a valuable use of my time to retake this one.

Having it there also provides an example of multiple status flags in the breadcrumbs in CMSMain which would otherwise not be shown. I don't want to add a custom one in this screenshot - that's what `status-flags-custom.png` is for.

Comment (ChloeSartorelli) [reply en/_images/page_node_normal.png:1]: > status-flags-custom.png has misaligned text in the badges

It's a screenshot of how it looks.

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md (hunk 5 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md]: As discussed in person since the problem you had was the word "published" because it's similar to versioning concepts I've just reworked it to avoid that word while still giving a meaningful example that hints at the sorts of use cases this might have.

I think using `_t()` in examples promotes best practice but as discussed will leave it out for the sake of getting this merged.

I've also updated the screenshot so it reflects the custom flag from the example.

File: en/08_Changelogs/6.0.0.md (hunk 4 above)
Comment (ChloeSartorelli) [reply en/08_Changelogs/6.0.0.md]: Updated

File: en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md (hunk 6 above)
Comment (ChloeSartorelli) [reply en/02_Developer_Guides/15_Customising_the_Admin_Interface/07_Status_Flags.md:84]: Done

**Review by emteknetnz — APPROVED** [review]
